src/lib/provable/dynamic-array.ts

```diff
@@ -0,0 +1,598 @@
+      let offset = new Field(i).sub(n);
+      this.array[i] = Provable.if(
+        new Field(i).lessThan(n),
+        this.innerType,
+        NULL,
+        this.getOrUnconstrained(offset)
+      );
+    }
+  }
+
+  /**
+   * @returns a  new DynamicArray instance with the same values as the current
+   */
+  copy(): this {
+    let newArr = new (<any>this.constructor)();
+    newArr.array = this.array.slice();
+    newArr.length = this.length;
+    // TODO: assert equality of the two arrays?
+    return newArr;
+  }
+
+  /**
+   * Creates a new dynamic array with the values of the current array from
+   * index `start` (included) to index `end` (excluded). If `start` is not
+   * provided, it defaults to 0. If `end` is not provided, it defaults to the
+   * length of the array.
+   *
+   * @param start
+   * @param end
+   * @returns
+   */
+  slice(start?: Field, end?: Field): DynamicArray<T, V> {
+    start ??= new Field(0);
+    end ??= this.length;
+    let sliced = this.copy();
+    sliced.shiftLeft(start);
+    sliced.pop(this.length.sub(end));
+    return sliced;
+  }
+
+  /**
+   * Concatenates the current array with another dynamic array, returning a new
+   * dynamic array with the values of both arrays. The capacity of the new array
+   * is the sum of the capacities of the two arrays.
+   *
+   * @param other
+   * @returns
+   */
+  concat(other: DynamicArray<T, V>): DynamicArray<T, V> {
+    let res = this.growCapacityTo(this.capacity + other.capacity);
+    let offset = new Field(0).sub(new Field(this.length));
+    for (let i = 0; i < res.capacity; i++) {
+      res.array[i] = Provable.if(
+        new Field(i).lessThan(this.length),
+        this.innerType,
+        this.getOrUnconstrained(new Field(i)),
+        other.getOrUnconstrained(offset)
+      );
+      offset = offset.add(new Field(1));
+    }
+    res.length = this.length.add(other.length);
+    return res;
+  }
+
+  /**
+   * Inserts a value at index i, shifting all elements after that position to
+   * the right by one. The length of the array is increased by one, which must
+   * result in less than or equal to the capacity.
+   *
+   * @param i
+   * @param value
+   */
+  insert(index: Field, value: T): void {
+    const right = this.slice(index, this.length);
+    this.increaseLengthBy(new Field(1));
+    this.set(index, value);
+    for (let i = 0; i < this.capacity; i++) {
+      let offset = new Field(i).sub(index).sub(new Field(1));
+      this.array[i] = Provable.if(
+        new Field(i).lessThanOrEqual(index),
+        this.innerType,
+        this.getOrUnconstrained(new Field(i)),
+        right.getOrUnconstrained(offset)
+      );
+    }
+  }
+
+  // TODO: when we can check .equals on arrays eleemnts we can implement:
+  // - includes
+
+  // cached variables to not duplicate constraints if we do something like
+  // array.get(i), array.set(i, ..) on the same index
+  _indexMasks: Map<Field, Bool[]> = new Map();
```

Comment:
if we want these `_` properties/method to be private, I would prefer if we could use the `private` keyword or the `#` syntax